examples/src/rescue_raps/mod.rs

```diff
@@ -0,0 +1,172 @@
+        debug!(
+            "Computed two permuted chains of {} Rescue hashes in {} ms",
+            chain_length,
+            now.elapsed().as_millis(),
+        );
+
+        RescueRapsExample {
+            options,
+            chain_length,
+            seeds,
+            permuted_seeds,
+            result,
+        }
+    }
+}
+
+// EXAMPLE IMPLEMENTATION
+// ================================================================================================
+
+impl Example for RescueRapsExample {
+    fn prove(&self) -> StarkProof {
+        // generate the execution trace
+        debug!(
+            "Generating proof for computing a chain of {} Rescue hashes\n\
+            ---------------------",
+            self.chain_length
+        );
+
+        // create a prover
+        let prover = RescueRapsProver::new(self.options.clone());
+
+        // generate the execution trace
+        let now = Instant::now();
+        let trace = prover.build_trace(&self.seeds, &self.permuted_seeds, self.result);
+        let trace_length = trace.length();
+        debug!(
+            "Generated execution trace of {} registers and 2^{} steps in {} ms",
+            trace.width(),
+            log2(trace_length),
+            now.elapsed().as_millis()
+        );
+
+        // generate the proof
+        prover.prove(trace).unwrap()
+    }
+
+    fn verify(&self, proof: StarkProof) -> Result<(), VerifierError> {
+        let pub_inputs = PublicInputs {
+            result: self.result,
+        };
+        winterfell::verify::<RescueRapsAir>(proof, pub_inputs)
+    }
+
+    fn verify_with_wrong_inputs(&self, proof: StarkProof) -> Result<(), VerifierError> {
+        let pub_inputs = PublicInputs {
+            result: [self.result[1], self.result[0]],
+        };
+        winterfell::verify::<RescueRapsAir>(proof, pub_inputs)
+    }
+}
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+fn merge(values: &[[BaseElement; 2]; 2], result: &mut [BaseElement]) {
+    let mut state = BaseElement::zeroed_vector(STATE_WIDTH);
+    state[..2].copy_from_slice(&values[0]);
+    state[2..].copy_from_slice(&values[1]);
+    for i in 0..NUM_HASH_ROUNDS {
+        rescue::apply_round(&mut state, i);
+    }
+    result.copy_from_slice(&state[..2]);
+}
+
+fn compute_permuted_hash_chains(
+    seeds: &[[BaseElement; 2]],
+    permuted_seeds: &[[BaseElement; 2]],
+) -> [[BaseElement; 2]; 2] {
+    let mut values = [BaseElement::ZERO; 2];
+    let mut permuted_values = [BaseElement::ZERO; 2];
+    let mut tmp = [BaseElement::ZERO; 2];
+    let mut permuted_tmp = [BaseElement::ZERO; 2];
+
+    // Start the hash chain
+    for (seed, pseed) in seeds.iter().zip(permuted_seeds) {
+        merge(&[values, *seed], &mut tmp);
+        merge(&[permuted_values, *pseed], &mut permuted_tmp);
+        values.copy_from_slice(&tmp);
+        permuted_values.copy_from_slice(&permuted_tmp);
```

Comment:
If we change to the `absorb` function as described in the other comment, this could be changed to something like:

```Rust
absorb(&mut state, seed);
absorb(&mut permuted_state, speed);
```

Where `state` and `permuted_state` would track the state of the hashchains.